RzArch and plugins

wargio opened the pull request:
<!-- Filling this template is mandatory -->

**Your checklist for this pull request**
- [ ] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [ ] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [ ] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

Initial proposal and conversion of all the RzAsm/RzAnalysis plugins towards the unified RzArch* structures

Partially addresses https://github.com/rizinorg/rizin/issues/4334

wargio committed:
Initial version of RzArch and its internal structures.

wargio committed:
Add IL

XVilka commented on librz/include/rz_arch.h:
```diff
@@ -16,12 +20,140 @@ extern "C" {
+} RzArchDetailAccess;
+
+typedef enum {
+	RZ_ARCH_DETAIL_MEMBER_REGISTER = 0, ///< Member is raw bytes
```
Comment is wrong

XVilka replied:
Same for other members

XVilka commented on librz/include/rz_arch.h:
```diff
@@ -16,12 +20,140 @@ extern "C" {
 
+typedef struct rz_arch_t {
+	RzPVector /*<RzArchPlugin*>*/ *plugins;
+	HtSP /*<char*, RzConfig*>*/ plugins_config;
```
Is it a plugin-specific configuration?

XVilka reviewed (changes requested):
One big thing is missing in here - a target description, e.g.
```c
typedef struct {
    char *arch;
    char *cpu;
    ut16 bits;
} RzArchTarget;
```
This is where we can pack also the "features", e.g. particular ISA features or anything like that. Maybe also have a separate `address_width` member, e.g. see function `rz_analysis_get_address_bits()`

Rot127 commented on librz/include/rz_arch.h:
```diff
@@ -16,12 +20,140 @@ extern "C" {
 RZ_LIB_VERSION_HEADER(rz_arch);
 
+typedef enum {
+	RZ_ARCH_XCODE_MEMBER_BYTES = 0, ///< Member is raw bytes
```
~~Maybe add an invalid value so we can initialize it with `RZ_NEW0`~~
As discussed in Mattermost, we should consider just passing an `RzBuffer`.

Rot127 commented on librz/include/rz_arch.h:
```diff
@@ -16,12 +20,140 @@ extern "C" {
+
+typedef struct rz_arch_xcode_t {
+	RzArchXCodeMember member; ///< Describes the content of array
+	void *array; ///< Array of variable size, containing one or more structures of type member.
```
Better a `(P)Vector` or `Set`? Because we get iterators and such for free.

wargio replied:
Yes pvec is a better choice.

Rot127 commented on librz/include/rz_arch.h:
```diff
@@ -16,12 +20,140 @@ extern "C" {
+		size_t register_id;
+		ut64 imm_unsigned;
+		st64 imm_signed;
+	};
```
Also needs a bitvector for arbitrary sized values.

Rot127 commented on librz/include/rz_arch.h:
```diff
@@ -16,12 +20,140 @@ extern "C" {
+	RzAnalysisSwitchOp *switch_op;
+	RzAnalysisHint hint;
+	RzAnalysisDataType datatype;
+} RzArchPacket;
```
If `RzArchPacket` is supposed to be the "instruction packet" of VLIW, please move the this content into an `RzArchInsn` and make `RzArchPacket` a struct which holds multiple `RzArchInsn` in a vector (so we allow for order in them).

Would propose to implement iterators for `RzArchPacket` to access the single instructions:

- `RzIterator *rz_arch_packet_insn()` - Over all instructions
- `RzIterator *rz_arch_packet_calls()` - Over call instructions
- `RzIterator *rz_arch_packet_jumps()` - Over jumps
- `RzIterator *rz_arch_packet_mem()` - Over memory referencing instructions.
- `RzIterator *rz_arch_packet_specific(RzAnalysisOpType *atype)`
...

And getters for values with certain properties:

- `RzIterator *rz_arch_packet_code_refs(..., RzInsnRefType *ref_type)`
- `RzIterator *rz_arch_packet_data_refs(..., RzInsnRefType *ref_type)`

```
typedef enum {
  REF_TYPE_DATA = 1, ///< All data references
  REF_TYPE_CODE = 2, ///< All code references
  // Now more detailed characteristics
  REF_TYPE_STACK, ///< ref is stack related
  REF_TYPE_HEAP, ///< ref is heap related
  REF_TYPE_GLOBAL,
  REF_TYPE_CACHE,
  // Deprecated
  REF_TYPE_PTR, ///< returns former RzAnalysisOp.ptr
  REF_TYPE_STACKPTR, ///< returns former RzAnalysisOp.stackptr
  REF_TYPE_JUMP, ///< returns former RzAnalysisOp.jump
...
} RzRefType
```

Some or all of the should also be accessible via `xcode()`. `RZ_XCODE_MEMBER_STACK_REFS`, `RZ_XCODE_MEMBER_JMP_TARGETS`, `RZ_XCODE_MEMBER_CALL_TARGES` etc.

Because if we keep it like this, we starting coding again on the wrong abstraction layer of RzAnalysisOp.

Rot127 replied:
Ouh wait. I am stupid.
The packet can contain both. A vector with instructions (for VLIW archs) and single `RzAnalysisOp`/`RzAsm`/`RzArchInsn`.

Single instruction focused archs use always the last one. VLIW can implement everything over the vector of instructions.
And we only allow access to the details like `references`, `jump/call targets`, `asm_text` over the `xcode()`.

But the Packet itself should never be made available to the core.
But the iterators from above can be implemented just for plugin internal stuff.

Rot127 commented on librz/include/rz_arch.h:
```diff
@@ -16,12 +20,140 @@ extern "C" {
+typedef struct rz_arch_detail_t {
+	RzArchDetailValue source[6];
+	RzArchDetailValue destination;
+} RzArchDetail;
```
I struggle with the term `Detail`. Does it mean detail of operands? Or other details of the instruction? It looks like all operands of an instruction?
Also vectors are needed here again. Because we cannot assume a single destination or a just 6 sources etc.

If it is supposed to mean "RzArchOperand", as in "operand of an instruction", we should probably introduce a new object for them as well. Because they can become quite complex as well (e.g. memory operands consist of one reg, one imm. Or one reg, one implicit operand etc).

wargio replied:
maybe operand makes more sense

Rot127 commented on librz/include/rz_arch.h:
```diff
@@ -16,12 +20,140 @@ extern "C" {
+    bool (*init)(RZ_NONNULL RzConfig *config); ///< Global constructor for the plugin to fill the configuration values.
+    bool (*fini)(); ///< Global destructor for the plugin
+	bool (*can_xcode_in)(RZ_NONNULL RzArchXCodeMember input); ///< Returns true if the plugin can support the given RzArchXCodeMember in input.
+	bool (*can_xcode_out)(RZ_NONNULL RzArchXCodeMember output); ///< Returns true if the plugin can support the given RzArchXCodeMember in ouput.
```
An `can_xcode_transform` would be nice. To test if the plugin supports a given `in/out` combination. Because `can_xcode_in() && can_xcode_out() == true` but `context_xcode(in, out)` could be unsupported.

wargio replied:
i don't think this is an issue because we can define the transform, for example:
`asm > details` could become internally (in rzarch not plugins) `asm > bytes > details`.

Rot127 reviewed (commented):
Added quite some stuff which probably shouldn't be done in the first implementation.
I think necessary is though:
- The `RzArchPacket`/`RzArchInsn` distinction.
- Starting enums always with `invalid = 0` to prevent false positives.

Rot127 commented:
If we allow to query for specific properties via `xcode()`, we should implement also a common buffer mechanism for disassembled instructions of a plugin.
So the plugin doesn't need to disassemble everything again and again, just because the code asks for `call refs`, `stack refs`, `asm_text` in a row.

Rot127 committed:
Replace Bytes member with RzBuffer member.

RzBuffer is more versatile.

Rot127 committed:
Ad proposal of RzArchPacket design.

Rot127 committed:
Change member description back to bytes.

The xcode function is actually from bytes -> asm/packet etc.
The RzBuffer is just the vehicle to provide the bytes.

Rot127 committed:
Change members to RzPVector

Rot127 committed:
Move and extend detail object to operand object.

Rot127 committed:
Add a grouped iterator

XVilka commented:
This needs to be updated on top of the latest `dev` because of the recent changes in RzAnalysis and RzAsm

Rot127 committed:
Add basic proposal for plugin storage/management.
